### PR TITLE
[3.12] gh-87891: Add ABI check to CI

### PR DESCRIPTION
**Pull request by encukou**
- Backport the workflow step and fix-ups:
    - GH-92442 (e89c01eac7731d7cb54d43252dbc3d3f3a040c53)
    - GH-94129 (0dadb2249a8bafa7c5877daa08c9452f2248958a)
    - GH-98556 (194588decc05fa12f04cd90c3b78cc081151b19e)
- Remove ABI dump from gitignore
- Add ABI dump generated in a ubuntu:22.04 container

cc @Yhg1s 

----
:books: Documentation preview :books:: https://cpython-previews--104793.org.readthedocs.build/

* Issue: gh-87891

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -75,6 +75,36 @@ jobs:
             echo "run_hypothesis=true" >> $GITHUB_OUTPUT
           fi
 
+  check_abi:
+    name: 'Check if the ABI has changed'
+    runs-on: ubuntu-20.04
```
Comment (hugovk) [inline .github/workflows/build.yml]: Can we use?
```suggestion
    runs-on: ubuntu-latest
```
Or:
```suggestion
    runs-on: ubuntu-22.04
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -75,6 +75,36 @@ jobs:
...
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -75,6 +75,36 @@ jobs:
...
+          CFLAGS: -g3 -O0
+        run: |
+          # Build Python with the libpython dynamic library
+          ./configure --enable-shared
+          make -j4
+      - name: Check for changes in the ABI
+        run: |
+          if ! make check-abidump; then
+            echo "Generated ABI file is not up to date."
+            echo "Please, add the release manager of this branch as a reviewer of this PR."
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
            echo "Please add the release manager of this branch as a reviewer of this PR."
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -75,6 +75,36 @@ jobs:
...
+            echo ""
+            echo "To learn more about this check, please visit: https://devguide.python.org/setup/?highlight=abi#regenerate-the-abi-dump"
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
            echo "To learn more about this check: https://devguide.python.org/setup/#regenerate-the-abi-dump"
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -75,6 +75,36 @@ jobs:
...
 
+  check_abi:
+    name: 'Check if the ABI has changed'
+    runs-on: ubuntu-20.04
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - name: Install Dependencies
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
      - name: Install dependencies
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -75,6 +75,36 @@ jobs:
             echo "run_hypothesis=true" >> $GITHUB_OUTPUT
           fi
 
+  check_abi:
+    name: 'Check if the ABI has changed'
```
Comment (hugovk) [inline .github/workflows/build.yml:79]:
```suggestion
    name: 'Check if ABI has changed'
```

File: .github/workflows/build.yml (hunk 1 above)
Comment (encukou) [reply .github/workflows/build.yml]: Not latest, we need to pin the version here.

File: .github/workflows/build.yml (hunk 6 above)
Comment (encukou) [reply .github/workflows/build.yml:79]: I'm not sure if this is used in the required checks list.

Comment (Yhg1s) [reply .github/workflows/build.yml:79]: It is. Also, "the ABI" is better (it is a noun, after all), so why change it?

Comment (hugovk) [reply .github/workflows/build.yml:79]: To make the text a bit shorter, especially helps for limited space mobile. But it doesn't matter, especially if it's a required check.

**Pull request merged**